blocks/datasets/__init__.py

```diff
@@ -26,51 +27,51 @@ class Dataset(object):
 
     """
     __metaclass__ = ABCMeta
-    sources = ('features', 'targets')
 
-    def __iter__(self):
-        """Use the default iteration scheme to construct a data stream.
-
-        .. warning::
-
-           A dataset only produces a single data stream using the default
-           iteration scheme, so multiple iterations in parallel are not
-           supported.
-
-        """
-        if not hasattr(self, 'default_scheme'):
-            raise NotImplementedError("Does not provide a default iterator")
-        elif not hasattr(self, 'default_stream'):
-            self.default_stream = DataStream(self, self.default_scheme)
-        return iter(self.default_stream)
-
-    def open(self, state=None):
+    @abstractmethod
+    def open(self):
         """Return the state if the dataset requires one.
 
         Datasets which e.g. read files from disks require open file
         handlers, and this sort of stateful information should be handled
         by the data stream.
 
-        Parameters
-        ----------
+        Returns
+        -------
         state : object
-            The state of the dataset after the previous epoch.
+            An object representing the state of a dataset.
+
+        """
+        pass
+
+    def reset(self, state):
+        """Resets the state.
+
+        The default implementation closes the state and opens a new one.
+        A more efficient implementation can override the default one
+        in descendant classes.
 
         Returns
         -------
         state : object
-            An object representing the state of a dataset.
+            A reseted state.
 
+        """
+        self.close(state)
```

Comment:
I consciously didn't do this because there might be faster ways than closing and opening to reset e.g. seek to the beginning of a file. Or if the dataset is an interface to a web API or a database, you don't want to redo the authentication process (but you might need to send some sort of reset signal).